text/0000-futures.md

```diff
@@ -0,0 +1,617 @@
+onto a single operating system thread.
+
+To perform this cooperative scheduling we use a technique sometimes referred to
+as a "trampoline". When a task would otherwise need to block waiting for some
+event, instead it saves an object that allows it to get scheduled again later
+and *returns* to the executor running it, which can then run another task.
+Subsequent wakeups place the task back on the executors queue of ready tasks,
+much like a thread scheduler in an operating system.
+
+Attempting to complete a task (or async value within it) is called *polling*,
+and always yields a `Poll` value back:
+
+```rust
+/// Indicates whether a value is available, or if the current task has been
+/// scheduled for later wake-up instead.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Poll<T> {
+    /// Represents that a value is immediately ready.
+    Ready(T),
+
+    /// Represents that a value is not ready yet.
+    ///
+    /// When a function returns `Pending`, the function *must* also
+    /// ensure that the current task is scheduled to be awoken when
+    /// progress can be made.
+    Pending,
+}
+```
+
+When a task returns `Poll::Ready`, the executor knows the task has completed and
+can be dropped.
+
+### Waking up
+
+If a future cannot be directly fulfilled during execution and returns `Pending`,
+it needs a way to later on inform the executor that it needs to get polled again
+to make progress.
+
+This functionality is provided through a set of `Waker` types.
+
+`Waker`s are objects which are passed as a parameter to the `Future::poll` call,
+and which can be stored by the implementation of those `Futures`s. Whenever a
+`Future` has the need to get polled again, it can use the `wake` method of the
+waker in order to inform the executor that the task which owns the `Future`
+should get scheduled and executed again.
+
+The RFC defines a concrete `Waker` type with which implementors of `Futures`
+and asynchronous functions will interact. This type defines a `wake(&self)`
+method which is used to schedule the task that is associated to the `Waker`
+to be polled again.
+
+The mechanism through which tasks get scheduled again depends on the executor
+which is driving the task.
+Possible ways of waking up an executor include:
+- If the executor is blocked on a condition variable, the condition variable
+  needs to get notified.
+- If the executor is blocked on a system call like `select`, it might need
+  to get woken up by a syscall like `write` to a pipe.
+- If the executor's thread is parked, the wakeup call needs to unpark it.
+
+To allow executors to implement custom wakeup behavior, the `Waker` type
+contains a type called `RawWaker`, which consists of a  pointer
+to a custom wakeable object and a reference to a virtual function
+pointer table (vtable) which provides functions to `clone`, `wake`, and
+`drop` the underlying wakeable object.
+
+This mechanism is chosen in favor of trait objects since it allows for more
```

Comment:
This paragraph is not obviously true to me. Could a specific example be provided showing how a raw vtable works better than a trait object? I can't think of a situation in which manually defining the vtable can unlock a capability vs. a trait object.

Specifically, this API seems to provide a significant forwards compatibility hazard. With traits, additional fns can be added w/o breaking changes by including a default implementation. I do not see how this can be accomplished given the raw vtable.

Reply:
> Additionally, this API seems to provide a significant forwards compatibility hazard. With traits, additional fns can be added w/o breaking changes by including a default implementation. I do not see how this can be accomplished given the raw vtable.

This can be solved by using `#[non_exhaustive]` as well as exposing a constructor function on `RawWaker` that takes the minimum required things (i.e. the things required today).

Reply:
@Centril 

While I agree the API **can** be made forwards compatible, it is still unclear to me *why* the raw vtable is being used instead of trait objects.

If the raw vtable is used, the RFC should be updated to:

- Include examples illustrating why the raw vtable is a superior option.
- Make the API forwards compatible.

Reply:
How would you write the trait object version of `RawWaker` without using `Box`?

Reply:
@jethrogb that entirely depends on the details of the executor.

Reply:
@carllerche I think @jethrogb's question was more along the lines of "how would you represent a trait object without using `Box`?" not about how the executor actually constructs a `RawWaker`. There has to be a way to represent a conceptually owned "Wake" object on `no_std` systems where `Box` isn't an option.

When you say "trait object," what exactly do you have in mind?

Reply:
@jrobsonchase @jethrogb Exactly what @seanmonstar said above.

You can have unsafe trait obj pointers. We've been using `*mut UnsafeWake` for years with no problems (that I know of). There are ways of achieving anything possible w/ manually defining the vtable (again, AFAIK).

Again, if there is a use case enabled by specifying a manual vtable, great. But it should be specified in the RFC.

Reply:
This one:

```rust
    /// FIXME(cramertj)
    /// This method is intended to have a signature such as:
    ///
    /// ```ignore (not-a-doctest)
    /// fn drop_raw(self: *mut Self);
    /// ```
    ///
    /// Unfortunately in Rust today that signature is not object safe.
    /// Nevertheless it's recommended to implement this function *as if* that
    /// were its signature. As such it is not safe to call on an invalid
    /// pointer, nor is the validity of the pointer guaranteed after this
    /// function returns.
```

Or is that no longer relevant and just failed to get removed?

Reply:
@jrobsonchase IMO that wart seems relatively minor in context.

The forwards compatibility issue is real IMO. @Centril called out future proofing this. I assume that would require the constructor to have const fns? And, every time there is a change, creating a new constructor?

Reply:
One other important conceptual difference is that the raw vtable doesn't imply that much that we are talking about a real object. E.g. current executors/wakeups are implement through various mechanisms:
- As global functions
- As member functions on refcounted objects
- As thread-local functions.

All of those could be modeled with `UnsafeWake` too, but the casts of that some things into that trait were not really a natural fit. It seemed like mostly people with good experience on how trait objects look like in memory could implement it.

> With traits, additional fns can be added w/o breaking changes by including a default implementation. I do not see how this can be accomplished given the raw vtable.

I don't see compatibility issues. Either the default implementation can be added through new constructors, or the relevant field on `RawWakerVTable` simply stays `null`, and `Waker` calls the default method when the thing is not populated.
In the C world this approach is even used to extend things in an ABI compatible way - which is something we even don't require at the moment.